cmip6_downscaling/methods/detrend.py

```diff
@@ -0,0 +1,104 @@
+import pandas as pd
+
+
+def add_circular_temporal_pad(data, offset, timeunit='D'):
+    """
+    pad the beginning of data with the last values of data, and pad the end of data with the first values of data
+
+    data must have a dimension called time
+    """
+
+    padded = data.pad(time=offset, mode='wrap')
+
+    time_coord = padded.time.values
+    time_coord[:offset] = (data.time[:offset] - pd.Timedelta(offset, timeunit)).values
+    time_coord[-offset:] = (data.time[-offset:] + pd.Timedelta(offset, timeunit)).values
+
+    padded = padded.assign_coords({'time': time_coord})
+
+    return padded
+
+
+def days_in_year(year, use_leap_year=False):
+    if (year % 4 == 0) and use_leap_year:
+        return 366
+    return 365
+    # there might be calendars with 360 days
+
+
+def pad_with_edge_year(data):
+    """
+    pad data with repeating values at the edge, similar to the behavior of np.pad(mode='edge') but uses the 365 edge values
+    instead of repeating only 1 edge value (366 if leap year)
+
+    offset is the number of years to pad on each side
+    data must have a dimension called time
+    """
+
+    def pad_with(vector, pad_width, iaxis, kwargs):
+        pstart = pad_width[0]
+        pend = pad_width[1]
+        if pstart > 0:
+            start = vector[pstart : pstart * 2]
+            vector[:pstart] = start
+        if pend > 0:
+            end = vector[-pend * 2 : -pend]
+            vector[-pend:] = end
+
+    # TODO: figure out whether use_leap_year should be true or false based on the
+    prev_year = days_in_year(data.time[0].dt.year.values - 1, use_leap_year=False)
+    next_year = days_in_year(data.time[-1].dt.year.values + 1, use_leap_year=False)
```

Comment:
You may find `DataArray.dt.is_leap_year` useful in this context: http://xarray.pydata.org/en/stable/generated/xarray.DataArray.dt.is_leap_year.html